### ISSUE TITLE
flakes: add OCI registry as an input

### ISSUE THREAD

**Issue by KFearsoff** [NixOS/nix#7180]
**Is your feature request related to a problem? Please describe.**
NixOS provides a module `virtualisation.oci-containers.containers`, which lets the user create declarative Docker containers. This is great for NixOS adoption, since Docker images are a standard way of distributing services nowadays, and managing them with NixOS lets people use services not yet packaged for NixOS.
However, `virtualisation.oci-containers.containers` option needs the user to specify a tag to pull, which means that container versions are not managed by NixOS: you'd have to specify their versions by hand, and if you use a generic tag instead of a concrete hash, your system won't be reproducible.

**Describe the solution you'd like**
An additional input type for Flakes:
```nix
{
  inputs.photoprism.url = "oci+https:docker.io/photoprism/photoprism?latest";
}
```
It would then be possible to refer to it like this:
```nix
virtualisation.oci-containers.containers.photoprism.image = inputs.photoprism;
```

**Describe alternatives you've considered**
1. Fetch the tarball instead of the image, use `inputs` type "tarball", use `imageFile` to import the tarball as the container. As far as I know, [fetching the tarballs isn't possible from Docker registries](https://devops.stackexchange.com/questions/2731/downloading-docker-images-from-docker-hub-without-using-docker).
2. Specify image tags explicitly, update them whenever you run `nix flake update`. It doesn't really scale when you have a lot of containers.
3. Pull "latest", write a convenience script that would fetch newer versions when `nix flake update` is ran. This still isn't reproducible.

**Additional context**
It looks like fetching the whole container as a dependency for a flake is a really bad idea. A better approach seems to be to fetch [OCI manifest](https://docs.docker.com/registry/spec/manifest-v2-2/), which provides the "digest" which is essentially a commit hash. Fetching a JSON is cheap enough, and having a "digest" is enough to pin a docker image. I'm not sure, but it looks like the [necessary API is there](https://docs.docker.com/registry/spec/api/#get-manifest) to fetch the manifest, which can then be parsed.

**Comment by SuperSandro2000** [comment]
> ```nix
> {
>   inputs.photoprism.url = "oci+https:docker.io/photoprism/photoprism?latest";
> }
> ```

You are sure this is not ``oci+https:docker.io/photoprism/photoprism:latest``?

**Comment by Tomaszal** [comment]
I agree, this seems like a natural nix way to lock OCI images. This and [Arion](https://github.com/hercules-ci/arion) would make spinning up and maintaining most self-hosted solutions extremely simple. Sadly, I've yet to find a ready solution for managing OCI versions in docker. I have, however, found [Hocker](https://github.com/awakesecurity/hocker), which seems to do half the job already. It would probably be possible to write a simple-ish wrapper script around it to create a `docker.lock` file of sorts, which could function similarly. Of course, having this built into flakes would be much nicer.

**Comment by GinnyGlider** [comment]
Yes.

This is the _elegant_ solution I was searching for, any update to this?

> "_[docker-lock](https://github.com/safe-waters/docker-lock) cli tool which solves this problem [...] this should do for now._"

That was about ~3 years ago, wonder what happened to the project. 
Looking at the [web archive](https://web.archive.org/web/20240905183644/https://github.com/safe-waters/docker-lock) seemed pretty good, and it was able to be used on other systems.

**Comment by srd424** [comment]
https://github.com/xyz-deploy-github/docker-lock seems to be a fork, also unmaintained .. if you extract the binary from the last container build at https://hub.docker.com/r/safewaters/docker-lock it seems to still work, it would be nice to see it brought back to life, or at least have the functionality added to something else!

**Comment by Ocramius** [comment]
Having automated `nix flake update ` upgrade upstream OCI references would be awesome.

I heavily rely on @renovatebot, and that already supports upgrading flake lock files.

That also means moving away from tools like watchtower, and towards reproducible docker image references 🤞

**Comment by MrNavaStar** [comment]
If we just need to grab the digest, we can use https://hub.docker.com/v2/repositories/photoprism/photoprism/tags/latest, and then we can pass it to the input variable as `photoprism@sha256:ff4c1e37e...`.

I am going to take a crack at implementing a custom fetcher for docker hub (and then maybe also one for ghcr.io if that goes well).
if you guys have more ideas or tips id love to hear them!

**Comment by ekursakov** [comment]
[npins](https://github.com/andir/npins) recently implemented container pinning using `nix-prefetch-docker` command, but it downloads the whole image to compute the image hash as it's required for `dockerTools.pullImage` and it's different from image digest (documentation: https://nixos.org/manual/nixpkgs/stable/#ssec-pkgs-dockerTools-pullImage-examples).
But if we don't use `pullImage` in input, and just provide image name string with digest, we can directly query the registry using for example skopeo tool (like `nix-prefetch-docker` already does):
```
skopeo inspect docker://ubuntu:latest --format '{{.Digest}}'
```
It'll support oci registry, so no need for specific fetchers.

Also, one more thing to consider is that OCI images for different platforms and architectures have different digests, so we'll need some way to specify it in the declaration or expose all of them as nested attributes (and probably map docker abis to nix ones).

**Comment by Ocramius** [comment]
> Also, one more thing to consider is that OCI images for different platforms and architectures have different digests, so we'll need some way to specify it in the declaration or expose all of them as nested attributes (and probably map docker abis to nix ones).

Would pointing at the main cross-platform manifest suffice? Or is that mutable?

**Comment by MrNavaStar** [comment]
Perhaps we can just pull down all the hashes for all architectures/platforms and pin them all? Essentially just grabbing everything at this url: https://hub.docker.com/v2/repositories/photoprism/photoprism/tags/latest

**Comment by MrNavaStar** [comment]
I suppose this only works for docker hub, I'm not sure if there is a cross registry api for all OCI registries.

**Comment by tcurdt** [comment]
> I suppose this only works for docker hub, I'm not sure if there is a cross registry api for all OCI registries.

You would need to specify which registry you are referring to - but the API should be a common one.
I am using this as a library for to resolve images.

https://github.com/regclient/regclient

**Comment by MrNavaStar** [comment]
Okay yeah I think as long as we declare the registry that could work. I think one of the endpoints here: https://specs.opencontainers.org/distribution-spec/?v=v1.0.0#endpoints would work.
